### PR TITLE
Readme - the high level description

### PR DESCRIPTION
**Pull request by tudor-malene**
### Why is this change needed?

We need a readme to explain what the project is, what it does, where to find various things, and how to run the project

### What changes were made as part of this PR:

Functional PR.
The high-level description o the major components.

A guide on how to start, and other useful bits will be added in subsequent PRs

### What are the key areas to look at
-

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,7 +1,154 @@
...
+![Architechture](design/obscuro_arch.jpeg)
+
+### I. The Enclave
+
+This is the core component of Obscuro which runs inside the TEE. 
+See [go/enclave](go/enclave)
+
+We use [EGo](https://www.edgeless.systems/products/ego/), an open source SDK for developing this confidential component.
+
+The Enclave exposes an [interface](go/common/enclave.go) over RPC which attempts to minimise the TCB (Trusted computing base).
```
Comment (GavT) [inline README.md:28]: Capitalisation of enclave is inconsistent

File: README.md [hunk 2]
```diff
@@ -1,7 +1,154 @@
...
+chains. To achieve this and minimise the effort and incompatibilities, we depend on [go-ethereum](https://github.com/ethereum/go-ethereum).
+
+The dependency on go-ethereum is not straight forward, since transaction execution is coupled with ethereum specific consensus rules,
+which had to be mocked out.
+
+See [go/enclave/evm](go/enclave/evm)
+
+
+#### 2. Store the state
+The dependency on go-ethereum for transaction execution, means that we have to use the same storage interfaces.
```
Comment (GavT) [inline README.md:43]: "we have to use the same" or just "we use the same". Have to sounds we're reluctant to use the same storage interfaces.

File: README.md [hunk 3]
```diff
@@ -1,7 +1,154 @@
...
+to be resistant to side-channel analysis which would allow an attacker to infer information on what calculations are being made
+based on what data the enclave is requesting or storing.
+
+In a future iteration, we'll look at alternatives to connect to a performant k/v store designed or modified for confidential computing.
+
+See [go/enclave/db](go/enclave/db)
+
+
+#### 3. Consume Ethereum blocks 
+The enclave is fed ethereum blocks through the rpc interface. These blocks are used as the "Source of Truth", and the enclave
```
Comment (GavT) [inline README.md:59]: Should ethereum be capitalised?

File: README.md [hunk 4]
```diff
@@ -1,7 +1,154 @@
...
+
+In the current iteration we use [EdglessDB](https://www.edgeless.systems/products/edgelessdb/), an open source database tailor-made 
+for confidential computing.
+
+Go-ethereum uses a key-value store interface, which we implemented on top of the SQL database. The reason for this odd choice 
+is that the data needs to be indexed and encrypted befored being sent for storage. And the operation of storing data needs 
+to be resistant to side-channel analysis which would allow an attacker to infer information on what calculations are being made
+based on what data the enclave is requesting or storing.
+
+In a future iteration, we'll look at alternatives to connect to a performant k/v store designed or modified for confidential computing.
```
[the remark this answers is not in the document]
Comment (GavT) [reply README.md:53]: Same for RPC since its an acronym

File: README.md [hunk 5]
```diff
@@ -1,7 +1,154 @@
...
+
+See [go/enclave/db](go/enclave/db)
+
+
+#### 3. Consume Ethereum blocks 
+The enclave is fed ethereum blocks through the rpc interface. These blocks are used as the "Source of Truth", and the enclave 
+extracts useful information from them, such as published rollups, deposits to the bridge, etc. Ethereum Re-orgs have to be detected
```
Comment (GavT) [inline README.md:60]: Re-orgs should not be capitalised. Even better, is there an alternative term which is more explanatory, like "Ethereum block reorganisation"

File: README.md [hunk 6]
```diff
@@ -1,7 +1,154 @@
...
+at this level to rollback the Obscuro state accordingly.
+
+To avoid the risk of the enclave being fed invalid blocks which an attacker can use to probe for information, or to shorten the
```
Comment (GavT) [inline README.md:63]: mentioning revelation period without explanation what it means could be confusing. Suggest not mentioning it at all here.

File: README.md [hunk 7]
```diff
@@ -1,7 +1,154 @@
...
+at this level to rollback the Obscuro state accordingly.
+
+To avoid the risk of the enclave being fed invalid blocks which an attacker can use to probe for information, or to shorten the 
+revelation period, the blocks have to be checked for validity, which includes checking that enough "work" went into them.
+To achieve this we depend on the [Blockchain](https://github.com/ethereum/go-ethereum/blob/e6fa102eb08c2b83ab75e85ca7860eea3a10dab0/core/blockchain.go) 
+logic.
+
+
+#### 4. Bridge to Ethereum 
+Obscuro is an Ethereum Layer 2, and one of the key aspects of layer2s is to feature a re-org resistent decentralised bridge.
```
Comment (GavT) [inline README.md:70]: "layer2s" should be "aspects of a Layer 2"
re-org should be written long hand

File: README.md [hunk 8]
```diff
@@ -1,7 +1,154 @@
...
+
+*Note that the current mempool implementation is very primitive. It always includes all received transactions that were not already
+included in a rollup.*
+
+
+#### 6. The rollups and the PoBI protocol
+
+Like in any blockchain the unit of the protocol is the batch of transactions organized in a chain. 
+The Obscuro blocks have an encrypted payload, which is only visible inside the secure enclave.
+All the logic of maintaining the current state based on incoming data and of producing new rollups is found in the
```
Comment (GavT) [inline README.md:104]: "All of the logic maintaining..."

File: README.md [hunk 9]
```diff
@@ -1,7 +1,154 @@
...
+
+
+### II. The Host
+
+The host service is the software that is under the control of the operator. It does not run inside a secure enclave, and there is no attestation on it.
+
+From a threat model point of view, the host service is seen as an adversary by an enclave. Any data that it feeds into the enclave
+will be verified and considered malicious.
+
+A secure solution that uses confidential computing will generally try to minimize the TCB, and run as much as possible outside the secure enclave,
```
Comment (GavT) [inline README.md:129]: Need to reference TCB earlier on if you're using the acronym

File: README.md [hunk 10]
```diff
@@ -1,7 +1,154 @@
...
+while still achieving the same security goals.
+
+The host service is the equivalent of a typical blockchain node, and is responsible for:
+ 
+- P2P messaging: Gossiping of encrypted transactions and rollups
+- RPC: Exposing an RPC interface similar to the one exposed by normal ethereum nodes
```
Comment (GavT) [inline README.md:135]: Capitalised Ethereum

File: README.md [hunk 11]
```diff
@@ -1,7 +1,154 @@
...
+- Communicating with an Ethereum node for retrieving blocks and for submitting transactions with data that was generated inside the enclave. 
+This means it has to be an ethereum wallet and control keys to accounts with enough Eth to publish transactions.
```
Comment (GavT) [inline README.md:137]: Ethereum and ETH.
This sentence does not read well. Perhaps "This means an Ethereum wallet and the control keys to accounts with enough ETH to publish transactions is required."

**Review by GavT — COMMENTED** [review]
Left some comments.
Also, do you plan to add a link to the whitepaper somewhere?

**Comment by tudor-malene** [comment]
> Left some comments. Also, do you plan to add a link to the whitepaper somewhere?

thanks. it's right in the first paragraph

File: README.md (hunk 2 above)
Comment (tudor-malene) [reply README.md:43]: freudian slip :)

File: README.md (hunk 5 above)
Comment (tudor-malene) [reply README.md:60]: re-org is a pretty widespread term. Whoever read this far, must be familiar with it

File: README.md (hunk 6 above)
Comment (tudor-malene) [reply README.md:63]: linking to the whitepaper

File: README.md [hunk 12]
```diff
@@ -1,30 +1,200 @@
...
+
+
+#### 2. Store the state
+The dependency on `go-ethereum` for transaction execution, means that we use the same storage interfaces.
+
+In the current iteration we use [EdglessDB](https://www.edgeless.systems/products/edgelessdb/), an open source database tailor-made 
+for confidential computing.
+
+`go-ethereum` uses a key-value store interface, which we implement on top of the SQL database. The reason for this odd choice 
+is that the data needs to be indexed and encrypted befored being sent for storage. And the operation of storing data needs
```
Comment (moraygrieve) [inline README.md:51]: Should read "before" rather than "befored"

File: README.md [hunk 13]
```diff
@@ -1,30 +1,237 @@
...
+![Architecture](design/obscuro_arch.jpeg)
+
+### I. The Enclave
+
+This is the core component of Obscuro which runs inside the TEE. 
+See [go/enclave](go/enclave)
+
+We use [EGo](https://www.edgeless.systems/products/ego/), an open source SDK for developing this confidential component.
+
+The Enclave exposes an [interface](go/common/enclave.go) over RPC which attempts to minimise the Trusted computing base(TCB).
```
Comment (moraygrieve) [inline README.md:30]: Should read Trusted "Computing Base (TCB)", or "trusted computing base (TCB)"?

File: README.md [hunk 14]
```diff
@@ -1,30 +1,237 @@
...
+at this level to rollback the Obscuro state accordingly.
+
+To avoid the risk of the Enclave being fed invalid blocks which an attacker can use to probe for information, or to shorten the 
+[revelation period](https://whitepaper.obscu.ro/obscuro-whitepaper/detailed-design.html#revelation-mechanism), the blocks have to be checked for validity, which includes checking that enough "work" went into them.
+To achieve this we depend on the [Blockchain](https://github.com/ethereum/go-ethereum/blob/e6fa102eb08c2b83ab75e85ca7860eea3a10dab0/core/blockchain.go) 
+logic.
+
+
+#### 4. Bridge to Ethereum 
+One of the key aspects of Ethereum Layer 2(L2) solutions is to feature a decentralised bridge that is resistant to 51% attacks.
```
Comment (moraygrieve) [inline README.md:72]: Space before the parenthesis? i.e. "Ethereum Layer 2 (L2)"

File: README.md [hunk 15]
```diff
@@ -1,30 +1,237 @@
...
+
+A secure solution that uses confidential computing will generally try to minimize the TCB, and run as much as possible outside the secure Enclave,
+while still achieving the same security goals.
+
+The Host service is the equivalent of a typical blockchain node, and is responsible for:
+ 
+- P2P messaging: Gossiping of encrypted transactions and rollups
+- RPC: Exposing an RPC interface similar to the one exposed by normal Ethereum nodes
+- Communicating with an Ethereum node for retrieving blocks and for submitting transactions with data that was generated inside the Enclave.
+  "This means an Ethereum wallet and the control keys to accounts with enough ETH to publish transactions is required.
```
Comment (moraygrieve) [inline README.md:164]: Remove double quote (there is no terminating one)?

File: README.md (hunk 13 above)
Comment (tudor-malene) [reply README.md:30]: the latter. I'll add quotes

**Review by BedrockSquirrel — APPROVED** [review]
LGTM as a general structure and intro the main components - we can fill it in and iterate on it over time

**Comment by tudor-malene** [comment]
> LGTM as a general structure and intro the main components - we can fill it in and iterate on it over time

thanks. yeah. perfect is the enemy of good.

**Pull request merged**